### ISSUE TITLE
Text placement within `div`

### ISSUE THREAD

**Issue by lpugin** [music-encoding/music-encoding#1189]
I am looking at encoding / rendering `div` for text in-between measures or systems as displayed in the example below. I imagine that text placement can be achieved with a `table`. It seems that using `div` within `div` would be simpler. However, as it stands, the position of a `div` cannot really be encoded.  Looking at existing attributes, having `@halign` and `@valign` on `div` would do the job. Then you can do:
```xml
<div>
  <div halign="left" valign="top">Some text at the top left of the div</div>
  <div halign="center" valign="middle">Text centered</div>
  <div halign="right" valign="bottom">Some text at the bottom right</div>
</div>
```
This would require adding these two attributes to `<div>`. Any objections?

Along the same lines, it would be meaningful to allow `<div>` within `pgHead` and `pgFoot` for positioning to be encoded in the same way.

![div-01-1](https://github.com/music-encoding/music-encoding/assets/689412/0d9fd074-a60a-4cfa-994f-cb8978bba94d)

**Comment by pe-ro** [comment]
The problem with this approach is that `<div>` doesn't allow textual content directly.  A possible solution is --
```xml
<div>
  <p halign="left" valign="top">Some text at the top left of the div</div>
  <p halign="center" valign="middle">Text centered</div>
  <p halign="right" valign="bottom">Some text at the bottom right</div>
</div>
```
The `<p>` element lacks `@halign` and `@valign`, but I don't think it's a problem to add them.  In fact, I suggest adding them to all block-level text elements:

- biblList
- castList
- div
- eventList
- fig (already has the attributes)
- lg
- list
- p
- quote
- sp
- table

If the semantics of `<p>` are a problem, then we could add a new `<ab>` (anonymous block) element, similar to the one in TEI (https://tei-c.org/release/doc/tei-p5-doc/en/html/ref-ab.html), and add the `@halign` and `@valign` attributes there, too.

Block-level text elements are already allowed within `<pgHead>` and `<pgFoot>`, so I wouldn't muddy the water by allowing `<div>` there.  `<p>` is already available, `<ab>` can be added there as well.

**Comment by pe-ro** [comment]
Sorry, the previous example should've been --
```xml
<div>
  <p halign="left" valign="top">Some text at the top left of the div</p>
  <p halign="center" valign="middle">Text centered</p>
  <p halign="right" valign="bottom">Some text at the bottom right</p>
</div>
```


Using anonymous block elements, the example would look like --
```xml
<div>
  <ab halign="left" valign="top">Some text at the top left of the div</ab>
  <ab halign="center" valign="middle">Text centered</ab>
  <ab halign="right" valign="bottom">Some text at the bottom right</ab>
</div>
```

**Comment by lpugin** [comment]
Yes, I realized that there was a problem with my proposal. I do find the semantics of `<p>` problematic and would prefer a more neutral element. At the same time, I find *anonymous block* a weird name. Why not `<tb>` for *text block*? I guess it depends how much we want to be aligned with TEI.

**Comment by pe-ro** [comment]
It's just a name, so `<tb>` would work, but I think it's important to be aligned with TEI when possible.

My understanding is that `@halign` and `@valign` position a block element bearing them within the context of its parent.  For instance, each of the `<ab>` elements in the example above is positioned within the parent `<div>` based on the `<ab>`'s `@halign` and `@valign` attributes. 

A similar thing happens with `<rend halign="left" valign="top">text here</rend>`, except that the context is inline text.

**Comment by pe-ro** [comment]
Just to make sure we're on the same page (ha ha),
```
<div>
  <ab halign="left" valign="top">
    <rend halign="right" valign="bottom">Some text</rend>
  </ab>
  <ab/>
  <ab/>
</div>
```
would position the first `<ab>` at the top left of its parent `<div>`, while the `<rend>` element would place its text at the bottom right of the `<ab>`.

It would seem then that attributes to describe the size and/or margins of the `<div>` and `<ab>` elements are also needed.

**Comment by lpugin** [comment]
Yes, I think we are. This makes sense. However, I think I would go one step further and **remove** `att.horizontalAlign` and `att.verticalAlign` from `rend`.  If you want to do what you described, you would use `ab` within `ab` (assuming you expect this to be possible:
```xml
<div>
  <ab halign="left" valign="top">
    <ab halign="right" valign="bottom">Some text</ab>
  </ab>
  <ab/>
  <ab/>
</div>
```
That way `rend` is exclusively about visual rendering and not about placement.

>  It would seem then that attributes to describe the size and/or margins of the `<div>` and `<ab>` elements are also needed.

Yes, you are right. However, if the blocks are not nested, it already work without it when doing automatic rendering by assuming the blocs would take all the space available / needed.

**Comment by pe-ro** [comment]
@lpugin, is this issue resolved now with respect to Verovio?  I take it that adding `<ab>` is no longer necessary unless you want to nest blocks, which `<p>` doesn't allow one to do.  But one can nest `<div>` elements and position their contents using `<p>`.

Where do we stand on adjusting the attributes of to `<div>` to account for size and margins?  Is that necessary at this point?

I'm not keen to remove `att.horizontalAlign` and `att.verticalAlign` from `<rend>` right now as they may be useful in situations other than block-level text; in figured bass, for example.

If your rendering questions are resolved, will you close this issue?

**Comment by lpugin** [comment]
Not quite. See https://github.com/music-encoding/music-encoding/issues/1189#issuecomment-1572054858 and a few more of our comments. I think we still need `att.horizontalAlign` and `att.verticalAlign` to be added to `<p>`, or to `<div>` but to `<p>` seems better to me. I don't think being able to represent nested blocks is really necessary.

**Comment by pe-ro** [comment]
@lpugin, Ah, I forgot about adding `att.horizontalAlign` and `att.verticalAlign`.  I'll make a PR to add them to all the block-level elements that don't already have them.

**Comment by lpugin** [comment]
ODD meeting: rename the `@func` attribute values into `system` or `measure` to make it clearer.
